### PR TITLE
Word graph hpp

### PR DESCRIPTION
**Pull request by james-d-mitchell**

### CODE REVIEW

**Comment by james-d-mitchell** [comment]
I have about 5 TODOs left to do and this is finished.

**Review by Joseph-Edwards — CHANGES_REQUESTED** [review]
Modulo my few brief comments, this looks good to me! I dragged my eyes over the documentation and the implementation in the `*.tpp` file to try and spot any obvious badness, but I haven't given it a thorough read. If there's a particular section that needs a bit more scrutiny I'm happy to have a more detailed look.

File: tests/test-paths.cpp [hunk 1]
```diff
@@ -60,11 +60,12 @@ namespace libsemigroups {
   struct LibsemigroupsException;  // forward decl
 
   namespace {
-    void add_chain(WordGraph<size_t>& digraph, size_t n) {
-      size_t old_nodes = digraph.number_of_nodes();
-      digraph.add_nodes(n);
-      for (size_t i = old_nodes; i < digraph.number_of_nodes() - 1; ++i) {
-        digraph.set_target(i, 0, i + 1);
+    // TODO add to word_graph helper namespace
```
[the remark this answers is not in the document]
Comment (james-d-mitchell) [reply tests/test-paths.cpp:63]: I want to keep this TODO, it's not yet done.

File: include/libsemigroups/word-graph.hpp [hunk 2]
```diff
@@ -1575,393 +2665,838 @@ namespace libsemigroups {
...
-  WordGraph<Node> to_word_graph(size_t num_nodes,
-                                std::initializer_list<std::vector<Node>> il);
+  [[nodiscard]] WordGraph<Node>
+  to_word_graph(size_t                                   num_nodes,
+                std::initializer_list<std::vector<Node>> targets);
 
+  //! \ingroup word_graph_group
+  //!
+  //! \copydoc to_word_graph(size_t,
+  //! std::initializer_list<std::vector<Node>> const&)
```
[the remark this answers is not in the document]
Comment (james-d-mitchell) [reply include/libsemigroups/word-graph.hpp:2710]: Annoyingly doesn't fit, but oh well

File: include/libsemigroups/word-graph.hpp [hunk 3]
```diff
@@ -129,63 +120,97 @@ namespace libsemigroups {
...
+    //! \param that the word graph to copy.
+    //!
+    //! \returns A reference to `*this`.
+    //!
+    //! \exceptions
+    //! \no_libsemigroups_except
+    //!
+    //! \par Complexity
+    //! \f$O(mn)\f$ where \p m is the number of nodes, and \p n is the
+    //! out-degree of the word graph.
```
[the remark this answers is not in the document]
Comment (james-d-mitchell) [reply include/libsemigroups/word-graph.hpp:208]: Well spotted.

File: tests/test-sims.cpp [hunk 4]
```diff
@@ -169,19 +169,29 @@ namespace libsemigroups {
...
-                      joiner.is_subrelation_no_checks(graphs[i], graphs[j]),
-                      graphs[i],
-                      graphs[j]));
+          REQUIRE(std::tuple(meeter.is_subrelation(graphs[i], graphs[j]),
+                             graphs[i],
+                             graphs[j])
+                  == std::tuple(joiner.is_subrelation(graphs[i], graphs[j]),
+                                graphs[i],
+                                graphs[j]));
+          // FIXME the below doesn't seem to work, but JDM expected it to
```
[the remark this answers is not in the document]
Comment (james-d-mitchell) [reply tests/test-sims.cpp:185]: Yup for later.

**Pull request merged**